### ISSUE TITLE
Diverging `modifies` clauses cause unsound vacuity

### ISSUE THREAD

**Issue by JustusAdam** [model-checking/kani#2909]
Diverging (panicking) expressions, such as indexing, in a `modifies` clause lead to vacuous verification.

**Comment by celinval** [comment]
Can you please clarify what you mean here? Maybe an example with the current output would be helpful. Will the overall verification succeed?

**Comment by jsalzbergedu** [comment]
An example of code that verifies vacuously, found with the help of @pi314mm :
```
// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

// Test that a modifies clause works when a (function call)
// expression is provided
fn infinite_looper() -> usize {
    loop {};
    1
}

#[kani::requires(arr[idx] == 1)]
#[kani::modifies(&arr[infinite_looper()])]
#[kani::ensures(arr[idx] == 2)]
fn modify(idx: usize, arr: &mut [i64]) {
    arr[idx] += 1;
}

#[kani::proof_for_contract(modify)]
fn main() {
    let mut arr: [i64; 4] = [1, 2, 3, 4];
    modify(1, &mut arr);
}
```

**Comment by zhassan-aws** [comment]
This can be reproduced without `modifies` clauses as well, e.g.:
```rust
#[kani::ensures(result == 1)]
fn foo() -> i32 {
    loop {};
    2
}

#[kani::proof_for_contract(foo)]
fn check_foo() {
    let _ = foo();
}
```

```
$ kani contract_loop.rs -Zfunction-contracts
Kani Rust Verifier 0.51.0 (standalone)
warning: unreachable expression
 --> contract_loop.rs:4:5
  |
3 |     loop {};
  |     ------- any code following this expression is unreachable
4 |     2
  |     ^ unreachable expression
  |
  = note: `#[warn(unreachable_code)]` on by default

warning: 1 warning emitted

warning: unreachable expression
 --> contract_loop.rs:4:5
  |
3 |     loop {};
  |     ------- any code following this expression is unreachable
4 |     2
  |     ^ unreachable expression
  |
  = note: `#[warn(unreachable_code)]` on by default

warning: Found the following unsupported constructs:
             - caller_location (1)
             - foreign function (2)
         
         Verification will fail if one or more of these constructs is reachable.
         See https://model-checking.github.io/kani/rust-feature-support.html for more details.

warning: 2 warnings emitted

Checking harness check_foo...
CBMC 5.95.1 (cbmc-5.95.1)
CBMC version 5.95.1 (cbmc-5.95.1) 64-bit x86_64 linux
Reading GOTO program from file /home/ubuntu/examples/contract_loop__RNvCsiqQmiVO2X69_13contract_loop9check_foo.out
...
SUMMARY:
 ** 0 of 1312 failed (1 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 1.7906353s

Complete - 1 successfully verified harnesses, 0 failures, 1 total.
```

**Comment by zhassan-aws** [comment]
And with a divergent `while` loop as well if loop contract synthesis is enabled:
```rust
#[kani::ensures(result == 1)]
fn foo() -> i32 {
    let i = 0;
    while i < 10 {}
    2
}

#[kani::proof_for_contract(foo)]
fn check_foo() {
    let _ = foo();
}
```

```bash
$ kani contract_loop.rs -Zfunction-contracts --synthesize-loop-contracts --enable-unstable
Kani Rust Verifier 0.51.0 (standalone)
warning: Found the following unsupported constructs:
             - caller_location (1)
             - foreign function (2)
         
         Verification will fail if one or more of these constructs is reachable.
         See https://model-checking.github.io/kani/rust-feature-support.html for more details.

warning: 1 warning emitted

Running loop contract synthesizer.
warning: This process may not terminate.
warning: Loop-contracts synthesizer is not compatible with unwinding bounds. Unwind bounds will be ignored.
Checking harness check_foo...
CBMC 5.95.1 (cbmc-5.95.1)
CBMC version 5.95.1 (cbmc-5.95.1) 64-bit x86_64 linux
Reading GOTO program from file /home/ubuntu/examples/contract_loop__RNvCsiqQmiVO2X69_13contract_loop9check_foo.out

...


SUMMARY:
 ** 0 of 1318 failed (1 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 1.8091615s

Complete - 1 successfully verified harnesses, 0 failures, 1 total.
```

**Comment by zhassan-aws** [comment]
One way to fix that is to fail if the post-condition is unreachable. In the case above, the check is indeed reported as unreachable:
```
Check 116: foo_check_516471.assertion.1
         - Status: UNREACHABLE
         - Description: "result == 1"
         - Location: contract_loop.rs:1:1 in function foo_check_516471
```

**Comment by celinval** [comment]
Why isn't this unwinding assertion failing in this case?

**Comment by zhassan-aws** [comment]
Apparently, CBMC replaces infinite loops with assume(false). This is an excerpt from the CBMC output:
```
Starting Bounded Model Checking
replacing self-loop at file contract_loop.rs line 5 column 5 function foo_wrapper_493980 by assume(FALSE)
no unwinding assertion will be generated for self-loop at file contract_loop.rs line 5 column 5 function foo_wrapper_493980
```

**Comment by zhassan-aws** [comment]
But in the case of using the loop synthesizer, one has to prove that the loop terminates for the loop abstraction to be sound.

**Comment by JustusAdam** [comment]
> Apparently, CBMC replaces infinite loops with assume(false). This is an excerpt from the CBMC output:
> 
> ```
> Starting Bounded Model Checking
> replacing self-loop at file contract_loop.rs line 5 column 5 function foo_wrapper_493980 by assume(FALSE)
> no unwinding assertion will be generated for self-loop at file contract_loop.rs line 5 column 5 function foo_wrapper_493980
> ```

This sounds like UB. C compilers also assume that you never write infinite loops.

**Comment by jsalzbergedu** [comment]
@JustusAdam ,
Could you provide the context in which this error can be reproduced? Building kani on your fork and then running it on this seems to fail.

**Comment by tautschnig** [comment]
#3223 should address most of what is discussed here, except https://github.com/model-checking/kani/issues/2909#issuecomment-2140906570 requires further investigation.

**Comment by JustusAdam** [comment]
Oh I made a mistake. I thought this issue was from the old `assigns` clause implementation, not `modifies` (I also observed vacuous verifications for that `assigns` clause though). I don't recall the example I tried this on, sorry.

**Comment by feliperodri** [comment]
I suggest that we close this issue once #3223 is merged and @qinheping could investigate the example from https://github.com/model-checking/kani/issues/2909#issuecomment-2140906570, which it seems to be more related to loop synthesis, as a separate issue. @zhassan-aws @tautschnig any thoughts?
